### DISCUSSION TITLE
Running instrumented XCTest on AWS Device Farm

### DISCUSSION THREAD

**Discussion by louwers** [bazelbuild/rules_apple#1993] [General]
I am building the test runner `ios_application` and submitting the `.ipa`. And separately I am building the `ios_unit_test` target and submitting the generated `.zip` with the XCTest.

It fails to run, getting this error:

```
	Reason: Cannot test target “RenderTest” on “PDX640618729”: Missing test product at /tmp/scratchEvkwZb.scratch/Productsh3Xt6n/Debug-iphoneos/RenderTestApp.app/PlugIns/RenderTest.xctest
```

Has anyone else gotten XCTests built with `rules_apple` to run on AWS Device Farm? 

https://docs.aws.amazon.com/devicefarm/latest/developerguide/test-types-ios-xctest.html

Maybe I need to enable "build for testing"?

Edit: It seems like they expect the .xctest to be part of the app.

**Comment by tinder-maxwellelliott** [comment]
I have successfully used [this logic](https://github.com/bazelbuild/rules_apple/pull/1994/files) to generate a xctestrun bundle that can run on real devices in Firebase test lab

**Reply by louwers** [reply]
Awesome, thanks for sharing. I will have a look.

**Reply by louwers** [reply]
I found a workaround, which is to use `rules_xcodeproj` with the `build_mode` set to `xcode`. And then use:

```
xcodebuild build-for-testing  -scheme RenderTest -archivePath $DIR -project MapLibre.xcodeproj
```

I wonder what the `rules_apple` equivalent of this would be.

**Reply by brentleyjones** [reply]
I wouldn't rely on `build_mode = "xcode"`, as it doesn't have first class support, and it might go away in the future. https://github.com/bazelbuild/rules_apple/pull/1994, when it lands, would be the rules_apple equivalent.

**Reply by louwers** [reply]
@brentleyjones We couldn't get our unit tests to work on device without it unfortunately.

**Reply by brentleyjones** [reply]
Then that's a bug. Can you file an issue? BwX shouldn't be required to have tests work.
